src/utils/daos_hdlr.c

```diff
@@ -1561,6 +1562,817 @@ cont_destroy_hdlr(struct cmd_args_s *ap)
+	sgl.sg_iovs = &iov;
+	sgl.sg_nr_out = 1;
+
+	rc = dfs_write(daos_file->dfs, daos_file->obj, &sgl, daos_file->offset, NULL); 
+
+	if (rc) {
+		fprintf(stderr, "dfs_write %s failed (%d %s)\n",
+			file, rc, strerror(rc));
+		errno = rc;
+		size = -1;
+	}
+
+	daos_file->offset += (daos_off_t)size;
+	return (ssize_t) size;
+}
+
+static ssize_t
+daos_file_write(daos_file_t *daos_file, char* file,
+		void* buf, size_t size)
+{
+	ssize_t num_bytes_written = 0;
+	if (daos_file->type == POSIX) {
+		num_bytes_written = write(daos_file->fd, buf, size);
+	} else if (daos_file->type == DAOS) {
+		num_bytes_written = daos_write(daos_file, file, buf, size);
+	} else {
+		fprintf(stderr, "File type not known: %s type=%d\n",
+			file, daos_file->type);
+	}
+	if (num_bytes_written < 0) {
+		fprintf(stderr, "write error on %s type=%d\n",
+			file, daos_file->type);
+	}
+	return num_bytes_written;
+}
+
+static void 
+daos_open(daos_file_t *daos_file, char *file, int flags, mode_t mode)
+{
+
+	int rc; 
+	dfs_obj_t *parent = NULL;
+	char *name = NULL, *dir_name = NULL;
+
+	parse_filename(file, &name, &dir_name);
+
+	assert(dir_name);
+
+	rc = dfs_lookup(daos_file->dfs, dir_name, O_RDWR, &parent, NULL, NULL);
+	if (parent == NULL) {
+		fprintf(stderr, "dfs_lookup %s failed \n", dir_name);
+	}
+
+	rc = dfs_open(daos_file->dfs, parent, name, mode | S_IFREG,
+		flags, 0, 0, NULL, &(daos_file->obj));
+	if (rc)
+		fprintf(stderr, "dfs_open %s failed (%d)\n", name, rc);
+}
+
+void
+daos_file_open(daos_file_t *daos_file, char* file, int flags, ...)
+{
+	/* extract the mode */
+	int mode_set = 0;
+	mode_t mode  = 0;
+	if (flags & O_CREAT) {
+		va_list ap;
+		va_start(ap, flags);
+		mode = va_arg(ap, mode_t);
+		va_end(ap);
+		mode_set = 1;
+	}
+
+	if (daos_file->type == POSIX) {
+		if (mode_set) {
+			daos_file->fd = open(file, flags, mode);
+		} else {
+			daos_file->fd = open(file, flags);
+		}
+	} else if (daos_file->type == DAOS) {
+		daos_open(daos_file, file, flags, mode);
+	} else {
+		fprintf(stderr, "File type not known: %s type=%d\n",
+			file, daos_file->type);
+	}
+}
+
+static int
+daos_mkdir(daos_file_t *daos_file, const char *path)
+{
+	int rc = 0;
+	dfs_obj_t *parent = NULL;
+	char *name = NULL;
+	char *dname = NULL;
+	parse_filename(path, &name, &dname);
+    
+	assert(dname);
+	if (name && strcmp(name, "/") != 0) {
+		rc = dfs_lookup(daos_file->dfs, dname, O_RDWR, &parent, NULL, NULL);
+		if (parent == NULL)
+			fprintf(stderr, "dfs_lookup %s failed \n", dname);
+		rc = dfs_mkdir(daos_file->dfs, parent, name, S_IRWXU, 0);
+		if (rc == EEXIST) {
+			fprintf(stdout, "dfs_mkdir %s already exists, %d \n", name, rc);
+		} else if (rc) {	
+			fprintf(stderr, "dfs_mkdir %s failed, %d \n", name, rc);
+		}
+	}
+	return rc;
+}
+
+static int 
+daos_file_mkdir(struct daos_file_t *daos_file, const char* dir, mode_t mode)
+{
+	int rc = 0;
+	if (daos_file->type == POSIX) {
+		rc = mkdir(dir, mode);
+	} else if (daos_file->type == DAOS) {
+		rc = daos_mkdir(daos_file, dir);
```

Comment:
you have the mode here from the input parameters, why are you not passing it to DFS?

Reply:
@mchaarawi updated, also added the chmod functions to update permissions to match the source file after copying